### DISCUSSION TITLE
HW4 Part 2,c

### DISCUSSION THREAD

**Discussion by gmg17005** [cooperrc/computational-mechanics#127] [General]
Does anyone know if we're supposed to pivot the matrix by hand before doing PLU decomposition? Or is there a function to be able to do that?

Thanks

**Comment by awc17002** [comment]
No need to do it by hand. In the reading there's two functions: one that returns the L and U matrices and the other that takes the L, U, and B matrix and returns the solutions. The functions are LUNaive and solveLU respectively.

**Reply by cooperrc** [reply]
Yes, you can use the `lu` method to find a good partial pivoting matrix, `p` because `p, l, u = lu(A)`, but make sure you use the pivot matrix on your right-hand-side `b` vector e.g. 

```python
AS = np.array([[0, 2, 3],
              [1, 5, 1],
              [2, 1, 1]])
print('AS=')
print(AS)

print('x = ',np.linalg.solve(AS, np.array([1,2,3])))

p, l, u = lu(AS)
print('LU=')
print(l@u)
print('x = ', np.linalg.solve(l@u, p@np.array([1,2,3])))
```
output:
```
AS=
[[0 2 3]
 [1 5 1]
 [2 1 1]]

x =  [1.32 0.08 0.28]

LU=
[[2. 1. 1.]
 [1. 5. 1.]
 [0. 2. 3.]]

x =  [1.32 0.08 0.28]
```
